File: python/02_system_prompt.py
#!/usr/bin/env python3
"""
Example 2: System + User Prompt
Control AI behavior with system instructions
"""

import http.client
import json
import os

api_key = os.environ.get('GROQ_API_KEY')
if not api_key:
    print("Error: GROQ_API_KEY not set")
    exit(1)

payload = {
    "model": "meta-llama/llama-4-scout-17b-16e-instruct",
    "messages": [
        {
            "role": "system",
            "content": "You are a concise expert in world history. Answer questions in exactly 2 sentences, no more."
        },
        {
            "role": "user",
            "content": "Why did the Roman Empire fall?"
        }
    ],
    "temperature": 0.5,
    "max_tokens": 150
}

conn = http.client.HTTPSConnection("api.groq.com")
headers = {
    'Content-Type': 'application/json',
    'Authorization': f'Bearer {api_key}'
}

conn.request("POST", "/openai/v1/chat/completions", json.dumps(payload), headers)
response = conn.getresponse()
response_data = json.loads(response.read().decode('utf-8'))
conn.close()

print("System Prompt Used:")
print(payload['messages'][0]['content'])

print("\nUser Question:")
print(payload['messages'][1]['content'])

print("\nAI Response (following system rules):")
print(response_data['choices'][0]['message']['content'])

print(f"\nToken Usage: {response_data['usage']['total_tokens']}")

# Key concepts:
#
# Message roles:
#   "system"    = Instructions for the AI's behavior
#   "user"      = Your questions/prompts
#   "assistant" = AI's previous responses (for conversation history)
#
# Message order:
#   1. System message (optional but recommended)
#   2. User message
#   3. Assistant message (if continuing conversation)
#   4. More user/assistant pairs for multi-turn chat
#
# System prompt examples:
#
#   Short responses:
#     "Be concise. Maximum 2 sentences."
#     "Answer in one word only."
#
#   Expertise level:
#     "You are a senior Python developer."
#     "Explain like I'm 5 years old."
#
#   Output format:
#     "Always respond with valid JSON."
#     "Use bullet points for all lists."
#     "Format code with markdown code blocks."
#
#   Personality:
#     "You are a friendly teacher."
#     "You are a formal business consultant."
#     "You are a pirate. Use pirate speak."
#
# Temperature explained:
#   0.0  = Deterministic (same answer every time)
#   0.3  = Focused, factual
#   0.5  = Balanced (good for facts that need slight variation)
#   0.7  = Creative but controlled
#   1.0  = Default (balanced creativity)
#   1.5  = Very creative
#   2.0  = Maximum randomness
#
# When to use different temperatures:
#   0.0-0.3: Math, facts, code generation
#   0.5-0.7: Explanations, summaries, Q&A
#   0.8-1.2: Creative writing, brainstorming
#   1.3-2.0: Poetry, experimental content
#
# Try this:
#   Change temperature to 0.0 and run multiple times
#   Change temperature to 2.0 and run multiple times
#   Compare the consistency of responses!

File: cpp/02_system_prompt.cpp
/**
 * ============================================================================
 * Example 2: System Prompt - Controlling AI Behavior
 * ============================================================================
 *
 * WHAT THIS DEMONSTRATES:
 *   - Using system prompts to set AI personality and behavior
 *   - The three message roles: system, user, assistant
 *   - How system messages influence AI responses
 *
 * WHAT YOU'LL LEARN:
 *   - System prompt patterns for different use cases
 *   - Message ordering and its importance
 *   - Creating specialized AI assistants
 *
 * COMPILE:
 *   g++ -std=c++17 -o 02_system_prompt 02_system_prompt.cpp -lcurl -I lib/
 *
 * RUN:
 *   ./02_system_prompt
 *
 * ============================================================================
 */

#include <iostream>
#include <string>
#include <cstdlib>
#include <curl/curl.h>
#include "json.hpp"

using json = nlohmann::json;

/* Callback for curl to write response data */
static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    std::string *response = static_cast<std::string*>(userp);
    response->append(static_cast<char*>(contents), size * nmemb);
    return size * nmemb;
}

int main() {
    /* Step 1: Get API key */
    const char *api_key_cstr = std::getenv("DEMETERICS_API_KEY");
    if (api_key_cstr == nullptr) {
        std::cerr << "Error: DEMETERICS_API_KEY not set\n";
        return 1;
    }
    std::string api_key(api_key_cstr);

    /* Step 2: Initialize curl */
    curl_global_init(CURL_GLOBAL_ALL);
    CURL *curl = curl_easy_init();
    if (curl == nullptr) {
        curl_global_cleanup();
        return 1;
    }

    /* ========================================================================
     * Step 3: Build the JSON request with a SYSTEM PROMPT
     * ========================================================================
     * The system prompt sets the AI's personality and rules.
     * It's like giving instructions to an employee before they start work.
     *
     * MESSAGE ROLES:
     *   - "system": Instructions for the AI (invisible to the "user")
     *   - "user": Messages from the human (you)
     *   - "assistant": Messages from the AI (in conversation history)
     */

    /* With nlohmann/json, building complex JSON is beautiful! */
    json request_body = {
        {"model", "meta-llama/llama-4-scout-17b-16e-instruct"},
        {"messages", {
            /* SYSTEM MESSAGE: Sets AI behavior - this AI will act as a pirate! */
            {
                {"role", "system"},
                {"content", "You are a friendly pirate who speaks in pirate dialect. "
                           "You always include nautical references and say 'Arrr!' occasionally. "
                           "Keep responses brief but entertaining."}
            },
            /* USER MESSAGE: The actual question */
            {
                {"role", "user"},
                {"content", "What is the capital of France?"}
            }
        }},
        {"temperature", 0.8},  /* Higher = more creative */
        {"max_tokens", 150}
    };

    std::string json_payload = request_body.dump();

    /* Step 4: Set up headers */
    struct curl_slist *headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");
    std::string auth_header = "Authorization: Bearer " + api_key;
    headers = curl_slist_append(headers, auth_header.c_str());

    /* Step 5: Configure and perform request */
    std::string response_data;
    curl_easy_setopt(curl, CURLOPT_URL, "https://api.demeterics.com/groq/v1/chat/completions");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_payload.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_data);

    std::cout << "Asking a pirate about France...\n\n";
    CURLcode res = curl_easy_perform(curl);

    if (res != CURLE_OK) {
        std::cerr << "Error: " << curl_easy_strerror(res) << "\n";
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
        curl_global_cleanup();
        return 1;
    }

    /* Step 6: Parse and display response */
    try {
        json response = json::parse(response_data);

        if (response.contains("choices") && !response["choices"].empty()) {
            std::string answer = response["choices"][0]["message"]["content"];
            std::cout << "Pirate AI says:\n" << answer << "\n";
        }
    } catch (const json::exception &e) {
        std::cerr << "Error parsing response: " << e.what() << "\n";
    }

    /* Step 7: Clean up */
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    curl_global_cleanup();

    /* Print examples of system prompts */
    std::cout << "\n";
    std::cout << "=== SYSTEM PROMPT EXAMPLES ===\n\n";

    std::cout << "1. EXPERT PERSONA:\n";
    std::cout << R"(   "You are a senior software engineer with 20 years experience.
    Give detailed technical explanations with code examples.")" << "\n\n";

    std::cout << "2. TEACHER:\n";
    std::cout << R"(   "You are a patient teacher for 10-year-old students.
    Use simple words and fun analogies.")" << "\n\n";

    std::cout << "3. JSON ONLY:\n";
    std::cout << R"(   "Respond only with valid JSON. No other text.
    Use the format: {\"answer\": \"...\"}")" << "\n\n";

    std::cout << "4. SAFETY:\n";
    std::cout << R"(   "You are a helpful assistant. Never provide harmful information.
    If asked about dangerous topics, politely decline.")" << "\n";

    return 0;
}

/*
 * SYSTEM PROMPT BEST PRACTICES:
 *
 * 1. Be specific about the persona
 *    Bad:  "Be helpful"
 *    Good: "You are a patient math tutor for high school students"
 *
 * 2. Define output format if needed
 *    "Always respond in JSON format"
 *    "Keep responses under 100 words"
 *
 * 3. Set boundaries
 *    "Only discuss topics related to cooking"
 *    "Do not provide medical advice"
 *
 * 4. Include examples
 *    "When greeting, say 'Howdy partner!' like a cowboy"
 *
 * 5. Order matters
 *    System message should always come FIRST
 *    Then conversation history (alternating user/assistant)
 *    Then the current user message
 *
 * C++ RAW STRING LITERALS:
 *   R"(...)" lets you write strings with quotes and newlines
 *   without escaping. Perfect for documentation!
 *
 *   R"(Hello "World")" is the same as "Hello \"World\""
 */
